source/PlayerInfo.cpp

```diff
@@ -2354,7 +2363,24 @@ void PlayerInfo::StepMissions(UI *ui)
 		else if(mission.CanComplete(*this))
 			RemoveMission(Mission::COMPLETE, mission, ui);
 		else if(mission.Destination() == GetPlanet() && !freshlyLoaded)
+		{
 			mission.Do(Mission::VISIT, *this, ui);
+			if(mission.IsUnique() || !mission.IsVisible())
+				continue;
+			
+			// On visit dialogs are handled separately as to avoid a player
+			// getting spammed by on visit dialogs if they are stacking jobs
+			// from the same destination.
+			if(visitText.empty())
+				visitText = Format::Replace(mission.GetAction(Mission::VISIT).DialogText(), substitutions);
+			++missionVisits;
+		}
+	}
+	if(!visitText.empty() && missionVisits > 1)
```

Comment:
Won't this mean a single visit dialog never gets shown? I think the guard needs to be around adding to the text.

Reply:
You're right.